File: src/solutions/algrithoms/Longest Increasing Subsequence/dp.c
#include <string.h>

#define max(a, b) (a > b ? a : b)

int lengthOfLIS(int* nums, int numsSize){
    int res = 0;
    int dp[numsSize];
    memset(dp, 0, numsSize * sizeof(int));
    for (int i = 0; i < numsSize; i++) {
        dp[i] = 1;
        for (int j = 0; j < i; j++) {
            if (nums[i] > nums[j]) {
                dp[i] = max(dp[i], dp[j] + 1);
            }
        }
        res = max(res, dp[i]);
    }
    return res;
}

File: src/solutions/algrithoms/House Robber II/dp.go
func rob(nums []int) int {
  n := len(nums)
  if n == 1 {
      return nums[0]
  }
  if n == 2 {
      return max(nums[0], nums[1])
  }
  return max(robCore(nums[1:]), robCore(nums[:n - 1]))
}

func robCore(nums []int) int {
  n := len(nums)
  dp := make([]int, n)
  dp[0] = nums[0]
  dp[1] = max(nums[0], nums[1])
  for i := 2; i < n; i++ {
      dp[i] = max(dp[i - 2] + nums[i], dp[i - 1])
  }
  return dp[n - 1]
}
